### ISSUE TITLE
`unnecessary_lazy_evaluations` fires when the closure has a `match`

### ISSUE THREAD

**Issue by ada4a** [rust-lang/rust-clippy#15832]
### Summary

I'd argue that `match` is not completely free, so it shouldn't be made eager? Though I guess it's a bit less expensive in my case, because the enum variants don't actually contain values that are then used in the match arms.

### Lint Name

unnecessary_lazy_evaluations

### Reproducer

I tried this code:

```rust
#![allow(unused)]
#![warn(clippy::unnecessary_lazy_evaluations)]

enum MyEnum {
    Foo,
    Bar,
    Baz,
}

fn foo(opt: Option<u32>, e: MyEnum) -> Option<u32> {
    opt.or_else(|| match e {
        MyEnum::Foo => Some(7),
        MyEnum::Bar => Some(13),
        MyEnum::Baz => None,
    })
}
```

I saw this happen:
```
warning: unnecessary closure used to substitute value for `Option::None`
  --> src/main.rs:11:5
   |
11 | /     opt.or_else(|| match e {
12 | |         MyEnum::Foo => Some(7),
13 | |         MyEnum::Bar => Some(13),
14 | |         MyEnum::Baz => None,
15 | |     })
   | |______^
   |
```

I expected to see this happen:
No lint

### Version

```text
rustc 1.92.0-nightly (3d8c1c1fc 2025-10-06)
binary: rustc
commit-hash: 3d8c1c1fc077d04658de63261d8ce2903546db13
commit-date: 2025-10-06
host: x86_64-unknown-linux-gnu
release: 1.92.0-nightly
LLVM version: 21.1.2
```

**Comment by arifd** [comment]
Purely aesthetic, but here's a simpler reproducer:
```rs
pub fn maybe(opt: Option<&str>) -> &str {
    opt
        .unwrap_or_else(|| match true {
            true => "true",
            false => "false",
        })
}
```

**Comment by arifd** [comment]
Interestingly, both compile to the same assembly: [godbolt](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:rust,selection:(endColumn:19,endLineNumber:13,positionColumn:19,positionLineNumber:13,selectionStartColumn:19,selectionStartLineNumber:13,startColumn:19,startLineNumber:13),source:'%23%5Binline(never)%5D%0Apub+fn+maybe_closure(opt:+Option%3C%26str%3E,+cond:+bool)+-%3E+%26str+%7B%0A++++opt%0A++++++++.unwrap_or_else(%7C%7C+match+cond+%7B%0A++++++++++++true+%3D%3E+%22true%22,%0A++++++++++++false+%3D%3E+%22false%22,%0A++++++++%7D)%0A%7D%0A%0A%23%5Binline(never)%5D%0Apub+fn+maybe_static(opt:+Option%3C%26str%3E,+cond:+bool)+-%3E+%26str+%7B%0A++++opt%0A++++++++.unwrap_or(match+cond+%7B%0A++++++++++++true+%3D%3E+%22true%22,%0A++++++++++++false+%3D%3E+%22false%22,%0A++++++++%7D)%0A%7D'),l:'5',n:'0',o:'Rust+source+%231',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:r1900,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1',verboseDemangling:'0'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:rust,libs:!(),options:'-C+opt-level%3D3',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+rustc+1.90.0+(Editor+%231)',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4
). Perhaps clippy got this one right.

**Comment by ada4a** [comment]
> both compile to the same assembly

The reviewer on the original code that motivated me to open this PR expressed a similar sentiment: https://github.com/rust-lang/rust-clippy/pull/15629#discussion_r2431201654

Maybe the diagnostic could incorporate this information somehow, otherwise the change sounds a bit unwarranted
